internal/generator/generator.go

```diff
@@ -59,6 +61,14 @@ func (puller ChartPuller) Pull(ctx context.Context, opts helm.PullOptions) error
 	mutex.Lock()
 	defer mutex.Unlock()
 
+	if _, err := os.Stat(opts.OutputDir); err == nil {
+		// If output directory exists it has been pulled by another goroutine
+		// No need to pull the chart
+		return nil
+	} else if !errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("failed to stat chart cache: %w", err)
+	}
+
```

Comment:
Isn't that a bit hacky and still subject to potential race conditions? Could you give more context on the choice of that strategy?

Reply:
So this is actually sound, and what we should have done last time, but I was too engrossed in the joy of writing concurrent code then thinking about exactly what needed to happen within the criticial section.

Essentially, multiple requests can happen in parallel, and so multiple releases that use the same chart can be evaluated in parallel. If a chart is not found in the cache, the cacheChart will attempt to invoke the helm CLI which will untar the chart into the output cache directory.

When we added the mutex we ensured that multiple goroutines that wanted to pull the same chart would do so sequentially, however, although they no longer pull in parallel, they will all eventually attempt to pull and will try to write untar the chart to the same directory. 

This causes the error.

By adding a check to the filesystem during the mutex lock, all goroutines that wish to pull a chart concurrently will do the following:
- grab the mutex for that chart
- check if the chart already exists and if so exits early
- pull the chart
- release the mutex

This way in a situation where multiple goroutines are trying to pull the same chart, the first to pull the chart will see that the directory does not exist and will download the chart and untar it. All following goroutines will see that the chart directory now exists and will opt out.

There is one slightly better implementation of this, which is to use `sync.Once` per chart pull, so that only one execution of the pull happens for every chart. However, although more efficient, it is harder to propagate errors, and you need to invalidate the `sync.Once` where errors occur so that the chart can be tried again.

So its a little more complicated to code, and is potentially more efficient at the cost of being more dangerous. Overall it's fine to check the cache directory a couple times, so its not worth it in my opinion.

/rant-over

Reply:
OK, got it! 👍

Actually, I think the approach with Mutex+dir-check is much better, because more stateless and tolerant to restarts, whereas I imagine that `sync.Once` would store its per-chart state in memory, which would not persist across restarts.

BTW: Are we handling error case where directory gets created, but chart pull fails? Are we able to retry on next attempt/call, if the condition is solely based on presence of directory?

Reply:
So the command that creates the directory is the `helm pull -untar` command which is not our code. We don't have anything to handle that at this time. 

This is an interesting edge case, but would need to be addressed within `joy`

Reply:
We cannot ensure that the chart is present within the directory, instead of only checking the mere presence of the directory?